src/sparse/impl/KokkosSparse_rcm_impl.hpp

```diff
@@ -0,0 +1,584 @@
+    IndexType sortBits = 0;
+    KeyType minKey = Kokkos::ArithTraits<KeyType>::max();
+    KeyType maxKey = 0;
+    Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(mem, n),
+    [=](size_type i, KeyType& lminkey)
+    {
+      if(keys[i] < lminkey)
+        lminkey = keys[i];
+    }, Kokkos::Min<KeyType>(minKey));
+    Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(mem, n),
+    [=](size_type i, KeyType& lmaxkey)
+    {
+      if(keys[i] > lmaxkey)
+        lmaxkey = keys[i];
+    }, Kokkos::Max<KeyType>(maxKey));
+    //apply a bias so that key range always starts at 0
+    //also invert key values here for a descending sort
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(mem, n),
+    [=](size_type i)
+    {
+      keys[i] -= minKey;
+    });
+    KeyType upperBound = maxKey - minKey;
+    while(upperBound)
+    {
+      upperBound >>= 1;
+      sortBits++;
+    }
+    for(IndexType s = 0; s < (sortBits + 3) / 4; s++)
+    {
+      //Count the number of elements in each bucket
+      IndexType count[16] = {0};
+      IndexType offset[17];
+      if(!inAux)
+      {
+        for(IndexType i = 0; i < n; i++)
+        {
+          count[(keys[i] & mask) >> maskPos]++;
+        }
+      }
+      else
+      {
+        for(IndexType i = 0; i < n; i++)
+        {
+          count[(keysAux[i] & mask) >> maskPos]++;
+        }
+      }
+      offset[0] = 0;
+      //get offset as the prefix sum for count
+      for(IndexType i = 0; i < 16; i++)
+      {
+        offset[i + 1] = offset[i] + count[i];
+      }
+      //now for each element in [lo, hi), move it to its offset in the other buffer
+      //this branch should be ok because whichBuf is the same on all threads
+      if(!inAux)
+      {
+        //copy from *Over to *Aux
+        for(IndexType i = 0; i < n; i++)
+        {
+          IndexType bucket = (keys[i] & mask) >> maskPos;
+          keysAux[offset[bucket + 1] - count[bucket]] = keys[i];
+          valuesAux[offset[bucket + 1] - count[bucket]] = values[i];
+          count[bucket]--;
+        }
+      }
+      else
+      {
+        //copy from *Aux to *Over
+        for(IndexType i = 0; i < n; i++)
+        {
+          IndexType bucket = (keysAux[i] & mask) >> maskPos;
+          keys[offset[bucket + 1] - count[bucket]] = keysAux[i];
+          values[offset[bucket + 1] - count[bucket]] = valuesAux[i];
+          count[bucket]--;
+        }
+      }
+      inAux = !inAux;
+      mask = mask << 4;
+      maskPos += 4;
+    }
+    //move keys/values back from aux if they are currently in aux,
+    //and remove bias
+    if(inAux)
+    {
+      Kokkos::parallel_for(Kokkos::ThreadVectorRange(mem, n),
+      [=](size_type i)
+      {
+        //TODO: when everything works, is safe to remove next line
+        //since keys (BFS visit scores) will never be needed again
+        keys[i] = keysAux[i];
+        values[i] = valuesAux[i];
+      });
+    }
+  }
+
+  //Functor that does breadth-first search on a sparse graph.
```

Comment:
It would be nice to expose BFS to the users as well. We could file an issue and come back later. No need to modify this PR.